CondFormats/CTPPSReadoutObjects/plugins/CTPPSPixelDAQMappingESSourceXML.cc

```diff
@@ -0,0 +1,585 @@
+
+  setWhatProduced(this, &CTPPSPixelDAQMappingESSourceXML::produceCTPPSPixelDAQMapping, es::Label(subSystemName));
+  findingRecord<CTPPSPixelDAQMappingRcd>();
+
+
+  LogVerbatim("CTPPSPixelDAQMappingESSourceXML") << " Inside  CTPPSPixelDAQMappingESSourceXML";
+
+
+}
+
+void CTPPSPixelDAQMappingESSourceXML::setIntervalFor(const edm::eventsetup::EventSetupRecordKey &key,
+						     const edm::IOVSyncValue& iosv, edm::ValidityInterval& oValidity)
+{
+  LogVerbatim("CTPPSPixelDAQMappingESSourceXML")
+    << ">> CTPPSPixelDAQMappingESSourceXML::setIntervalFor(" << key.name() << ")";
+
+  LogVerbatim("CTPPSPixelDAQMappingESSourceXML")
+    << "    run=" << iosv.eventID().run() << ", event=" << iosv.eventID().event();
+
+  currentBlockValid = false;
+  for (unsigned int idx = 0; idx < configuration.size(); ++idx)
+    {
+      const auto &bl = configuration[idx];
+
+      EventID startEventID = bl.validityRange.startEventID();
+      if (startEventID == EventID(1, 0, 1))
+	startEventID = EventID(1, 0, 0);
+
+      if (startEventID <= iosv.eventID() && iosv.eventID() <= bl.validityRange.endEventID())
+	{
+	  currentBlockValid = true;
+	  currentBlock = idx;
+
+	  const IOVSyncValue begin(startEventID);
+	  const IOVSyncValue end(bl.validityRange.endEventID());
+	  oValidity = ValidityInterval(begin, end);
+
+	  LogVerbatim("CTPPSPixelDAQMappingESSourceXML")
+	    << "    block found: index=" << currentBlock
+	    << ", interval=(" << startEventID << " - " << bl.validityRange.endEventID() << ")";
+
+	  return;
+	}
+    }
+
+  if (!currentBlockValid)
+    {
+      throw cms::Exception("CTPPSPixelDAQMappingESSourceXML::setIntervalFor") <<
+	"No configuration for event " << iosv.eventID();
+    }
+}
+
+
+
+CTPPSPixelDAQMappingESSourceXML::~CTPPSPixelDAQMappingESSourceXML()
+{
+}
+
+
+
+string CTPPSPixelDAQMappingESSourceXML::CompleteFileName(const string &fn)
+{
+  FileInPath fip(fn);
+  return fip.fullPath();
+}
+
+
+
+std::unique_ptr<CTPPSPixelDAQMapping> CTPPSPixelDAQMappingESSourceXML::produceCTPPSPixelDAQMapping( const CTPPSPixelDAQMappingRcd & )
+{
+  assert(currentBlockValid);
+
+  auto mapping = std::make_unique<CTPPSPixelDAQMapping>();
+  auto mask = std::make_unique<CTPPSPixelAnalysisMask>();
+
+// initialize Xerces
+  try
+    {
+      XMLPlatformUtils::Initialize();
+    }
+  catch (const XMLException& toCatch)
+    {
+      char* message = XMLString::transcode(toCatch.getMessage());
+      throw cms::Exception("CTPPSPixelDAQMappingESSourceXML") << "An XMLException caught with message: " << message << ".\n";
+      XMLString::release(&message);
+    }
+
+// load mapping files
+  for (const auto &fn : configuration[currentBlock].mappingFileNames)
+    ParseXML(pMapping, CompleteFileName(fn), mapping, mask);
+
+// load mask files
+  for (const auto &fn : configuration[currentBlock].maskFileNames)
+    ParseXML(pMask, CompleteFileName(fn), mapping, mask);
+
+// release Xerces
+  XMLPlatformUtils::Terminate();
```

Comment:
And here the corresponding method would be: `cms::concurrency::xercesTerminate();`